common/core/web/keyboard-processor/src/keyboards/activeLayout.ts

```diff
@@ -216,6 +216,26 @@ namespace com.keyman.keyboards {
       // This part depends on the keyboard processor's active state.
       if(keyboardProcessor) {
         keyboardProcessor.setSyntheticEventDefaults(Lkc);
+
+        // If it's a state key modifier, trigger its effects as part of the
+        // keystroke.
+        let bitmask = 0;
+        switch(Lkc.kName) {
+          case 'K_CAPS':
+            bitmask = text.Codes.stateBitmasks.CAPS;
+            break;
+          case 'K_NUMLOCK':
+            bitmask = text.Codes.stateBitmasks.NUM_LOCK;
+            break;
+          case 'K_SCROLL':
+            bitmask = text.Codes.stateBitmasks.SCROLL_LOCK;
+            break;
+        }
+
+        if(bitmask) {
+          Lkc.Lstates ^= bitmask;
+          Lkc.LmodifierChange = true;
+        }
```

Comment:
The replacement for the old, within-OSK approach.  This also correlates far more closely with what physical keystrokes do; they note the change in state, while the keyboard processor notices that change (hence `LmodifierChange`) and signals the need for an update.